### PR TITLE
feat(pubsub): Add Subscription#remove_dead_letter_policy

### PR DESCRIPTION
**Pull request by quartzmo**
This is an alternative solution to #6314 that has the advantage of not exposing a value object. The value object in #6314 may be in a state that is inconsistent with the service, due to the replacement of the special values `nil` and `0` with a default value.

closes: #6159

### CODE REVIEW

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb:435]: Seems to be some confusion about what `nil` means: whether it is because no dead letter policy is configured, or a default value is requested.

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb:518]: I would group these differently: first group all the "before" state together, then show the call, and group all the "after" state together:

```ruby
        #   sub.dead_letter_policy? #=> true
        #   sub.dead_letter_topic.name #=> "projects/my-project/topics/my-dead-letter-topic"
        #   sub.dead_letter_max_delivery_attempts #=> 10
        #
        #   sub.remove_dead_letter_policy!
        #
        #   sub.dead_letter_policy? #=> false
        #   sub.dead_letter_topic #=> nil
        #   sub.dead_letter_max_delivery_attempts #=> nil
```

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb:462]: Clarify: "The dead letter topic must be set *first*"

Comment (quartzmo) [reply google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb]: Thanks, updated.

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb
Comment (quartzmo) [reply google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb:517]: Good call, I agree. Updated.

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb
Comment (quartzmo) [reply google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb]: Removed the second sentence.

**Comment by quartzmo** [comment]
> I do think I like this better.

What do you think about adding similar methods for Retry Policy? And deprecating the methods recently added in #5944?

**Comment by dazuma** [comment]
@quartzmo Hmm, that's a good question. Maybe before we merge this, we should figure out how RetryPolicy would work, and make sure it does work. And see if that informs what the interface for dead letter policy should look like and how consistent we think these need to be.

In the case of RetryPolicy, I think the question is: what's the interface for moving from the "RetryPolicy absent" state to "RetryPolicy present but empty" state. We can't just set one of the fields, because they all remain `nil`. Would we need an `add_retry_policy!` method in addition to `remove_retry_policy!`?

**Comment by quartzmo** [comment]
Do you think it would be acceptable for users to be required to just explicitly set the two default values instead?

```
subscription.set_retry_policy_minimum_backoff = 10
subscription.set_retry_policy_maximum_backoff = 600
```

**Comment by dazuma** [comment]
No, I don't think so. That would require the user to know what those values are and hard-code them (which would also be a problem if the defaults ever changed on the backend).

**Comment by quartzmo** [comment]
I agree, there is a good use case for automatically migrating to updated defaults. I am OK with `#add_retry_policy!`

**Comment by dazuma** [comment]
For consistency, here's what I suggest:

* We provide `remove_<object>!` as implemented here, for both dead letter and retry policies. It takes no arguments, and simply removes the policy (and returns `true` if the policy was removed, or `false` if the policy was not there to begin with.)
* We also provide `add_<object>!` for _both_ dead letter and retry policies. This method takes keyword arguments allowing the user to set the fields of the policy when adding. In the retry policy case, both arguments are optional, defaulting to `nil` (which indicates the backend's default value). In the dead letter policy case, the `topic` argument is _required_ but `max_delivery_attempts` is optional and defaults to `nil`.
    * If the policy is absent when this is called, the policy is added with the given fields.
    * If the policy is _already present_ when this is called, then the fields of the existing policy are updated—the semantics should be as if the policy were deleted and then recreated using the given arguments.
* We keep the current "flattened" field accessors. Setting one of these fields (even to `nil`) implicitly adds the object. For retry policies, either field can be set at any time. For dead letter policies, it is an ArgumentError to set `dead_letter_topic` to `nil`. It is also an ArgumentError to set `dead_letter_max_delivery_attempts` if the object has not already been added.

**Comment by quartzmo** [comment]
We already have something kind of like what you are proposing in [`Subscription#policy`](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb#L1099-L1140). The IAM Policy is different in that I believe it always implicitly exists (and its existence has no impact unless `roles` is modified). However, there may be something in its design that we want to use, such as the block to set attributes. Note that although the `Policy` value object is public and exposed (unfrozen instance returned by `#policy`, changes persisted by [`Subscription#update_policy`](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb#L1142-L1176)), I'm pretty sure I recall that @blowmage regretted this decision and was hoping to migrate usage to the block-based update.

**Comment by quartzmo** [comment]
Here is the original [commit](https://github.com/googleapis/google-cloud-ruby/pull/192/commits/ee7597c2be39972be830c89bf178db20e06063cc) exposing the IAM Policy as a hash via getter/setter. It evolved from there to a value object and finally to being updated in the block to `#policy`.

**Comment by quartzmo** [comment]
Actually, I think [`Subscription#push_config`](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb#L223-L278) is a more completely evolved version of the IAM Policy design: The public value object [`PushConfig`](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/subscription/push_config.rb) is returned by the method, but unlike `Policy` it is frozen and there is no update method. The only way to update it is in the block to `#push_config`. This feature was added relatively recently in #3091, the original "flattened" access to `endpoint` is much older.

**Comment by quartzmo** [comment]
Another existing feature example: `Subscription#expires_in` was added in #3086, with this explanation of the flattening and a plan to later expose a value object:

> I decided to expose `expires_in` instead of an ExpirationPolicy because there is only the one field on ExpirationPolicy. If/when ExpirationPolicy adds more fields we can expose this structure while still allowing `expires_in` to directly access that value.

**Comment by quartzmo** [comment]
Adding `do not merge` until we settle on a common approach for all of these lower-level value objects.

**Review by dazuma — APPROVED** [review]

**Pull request merged**
